translate/v23tov30/v23tov30.go

```diff
@@ -231,6 +231,36 @@ func translateGroups(groups []old.PasswdGroup) (ret []types.PasswdGroup) {
 	return
 }
 
+func replaceOrAppendUnit(units []types.Unit, unit types.Unit) []types.Unit {
+	for i, u := range units {
+		if u.Name == unit.Name {
+			// Replace existing contents only if new contents are not empty.
+			// The unit may define dropins only and have no content,
+			// in which case we don't want to overwrite the unit contents
+			// and only append the dropins.
+			if unit.Contents != nil && unit.Contents != util.StrPStrict("") {
+				units[i].Contents = unit.Contents
+			}
+			if len(unit.Dropins) > 0 {
+				for _, newDropin := range unit.Dropins {
+					for j, existingDropin := range units[i].Dropins {
+						// replace dropin if already defined
+						if existingDropin.Name == newDropin.Name {
+							units[i].Dropins[j] = newDropin
+							continue
+						}
+						// else append the dropin
+						units[i].Dropins = append(units[i].Dropins, newDropin)
+					}
+				}
+			}
+			return units
+		}
+	}
+	units = append(units, unit)
```

Comment:
Aren't we mutating `units` here?

Reply:
Hm. It's the same structure as in the append files func.

Reply:
the dropin merge logic is still flawed here. I'll rework this.

Reply:
ok, I think this works better now.